### PR TITLE
Refactoring/Linting

### PR DESCRIPTION
**Pull request by brettz9**
- Refactoring: Use ES6 classes, destructuring; use `forEach`; declare
    variables in more nested scope
- Linting: Remove jshint
- Linting: Apply overrides to add mocha-specific config
- Linting: Add recommended fields to `package.json`
- Linting: Remove unused disable directives
- Git/npm/Linting: ignore `var`

#### Is this a security patch?

No.

### Checklist

- [x] I have read the [CONTRIBUTING](https://github.com/passport-next/passport/blob/master/CONTRIBUTING.md) guidelines. **`make view-cov` does not work, however, as I get the message, "The file /Users/brett/passport/reports/coverage/lcov-report/index.html does not exist."**
- [x] I have added test cases which verify the correct operation of this feature or patch. **N/A (still passing)**
- [x] I have added documentation pertaining to this feature or patch. **N/A: No new features**
- [x] The automated test suite (`$ make test`) executes successfully.
- [x] The automated code linting (`$ npm run-script lint`) executes successfully.

### CODE REVIEW

**Comment by brettz9** [comment]
By the way, my own ESLint config (at https://github.com/brettz9/eslint-config-ash-nazg ) incorporates even more useful linting than airbnb if you were open to trying it (it expands on "eslint:recommended" and "standard" and several other plugins/configs), though for stylistic rules, the current style could be enforced.

**Comment by adamhathcock** [comment]
Definitely happy using more ES6 features.

**Comment by brettz9** [comment]
Ok, I've pushed an additional commit which adds ash-nazg eslint config in place of airbnb. In order to demonstrate which airbnb rules we needed to add back (without overriding what I think are some useful stricter ash-nazg rules which airbnb disables), I am not currently extending from airbnb but have manually added its different (mostly) stylistic rules which the current code was expecting.

If you find this acceptable, I can remove the airbnb dev dependency.

If you prefer not to spell out the rules added back from airbnb manually and keep it as a dev dep and extend its config as well as ash nazg's, I can do that instead.

**Comment by brettz9** [comment]
As far as the failing travis build, this is due to the linting step which fails for pre-8.3 Node (due to a dev dependency on eslint-plugin-node using object rest). Wondering if we could at least skip the linting for this version, if support is still needed?

I'm not clear on why the coverage is failing, as I haven't really added code or removed testing.

**Comment by rwky** [comment]
I've dropped support for node 6 in the master branch since it's EOL and pushed to this branch, travis is now happy.

Leave this with me and I'll review the changes, if @adamhathcock or any other @passport-next/developers want to review it then please go ahead the more the merrier.

@brettz9 thanks for the contribution! Much appreciated :)

**Comment by brettz9** [comment]
Great, thanks @rwky ... I'm working on a PR now for a Promise-based API (currently `serializeUser`, `deserializeUser`, and `transformAuthInfo`). I'm also including an option to return values synchronously where the former two had not previously supported this (though it doesn't work with "pass" or "undefined" which would be ambiguous). I can remove that if that is not favored for some reason.

File: package.json [hunk 1]
```diff
@@ -28,24 +32,38 @@
...
+    "eslint-plugin-markdown": "^1.0.0",
+    "eslint-plugin-no-use-extend-native": "^0.4.0",
+    "eslint-plugin-node": "^9.0.1",
+    "eslint-plugin-promise": "^4.1.1",
+    "eslint-plugin-standard": "^4.0.0",
+    "eslint-plugin-unicorn": "^8.0.2",
     "make-node": "^0.4.6",
-    "mocha": "6.x.x"
+    "mocha": "6.x.x",
+    "typescript": "^3.4.5"
```
Comment (adamhathcock) [inline package.json:56]: typescript used?

**Review by adamhathcock — CHANGES_REQUESTED** [review]
Looks mostly fine to me.  Can't tell if it will break things but a major version bump seems sensible.

File: package.json (hunk 1 above)
Comment (brettz9) [reply package.json:56]: Just a peer dependency of one of the dependents of eslint-config-ash-nazg so added to devDeps to stop complaints of missing peer deps.

**Comment by adamhathcock** [comment]
In general, I'm scared of moving too far from upstream code.  I guess the question is how much do we want to risk not getting upstream changes/fixes?  I prefer Promises, obviously :)

**Comment by brettz9** [comment]
Isn't that the point of this repo though, that the original author wasn't responding? Since there haven't been any changes of note since then, I could submit these changes as a PR... I'd think ES features would be welcomed...

**Comment by adamhathcock** [comment]
I'm not against it.  I'm just making sure the decision is a measured one.

**Comment by rwky** [comment]
I'm up for promises, this repo is so far from the upstream repo that any fixes that do land upstream that I haven't already merged in would need to be manually applied assuming they're relevant to us.

If possible I'd rather have changes me semver minor instead of semver major i.e. new stuff doesn't break existing features. Promises are discussed here https://github.com/passport-next/passport/issues/7#issuecomment-421174501

**Comment by adamhathcock** [comment]
Let's start breaking things.

**Review by adamhathcock — APPROVED** [review]

**Comment by rwky** [comment]
I've checked the changes and I like them! I do want to make a few changes myself I'll sort it all out this week and merge this.

**Comment by rwky** [comment]
Just letting everyone know I've not forgotten about this just taking longer for me to do what I want than expected.

**Comment by brettz9** [comment]
No worries... Also, my apologies for not yet getting to the Promise PR which I still hope and plan to complete...

**Comment by brettz9** [comment]
I've gotten out the Promises changes in #21, though I'll probably want to try them out a bit in my own code to get a better feel for how it works in the real world. However, I'm wondering if you have an ETA on your changes desired for this Refactoring PR?

**Comment by rwky** [comment]
I've just got back off holiday so have the usual billion emails everyone gets when they get back ;) though this should be doable this week, I've got the code ready locally just need to tweak/test some things.

**Comment by rwky** [comment]
@brettz9 I've added a new PR for the refactoring in #22 can you take a look? If it's all ok and gets merged in then we can look at promises based off the new PR (it shouldn't take much effort to rebase).

**Pull request merged**